File: ios/Classes/SwahiliNFCPlugin.swift
import Flutter
import UIKit
import CoreNFC

@available(iOS 13.0, *)
public class SwahiliNFCPlugin: NSObject, FlutterPlugin, NFCNDEFReaderSessionDelegate {
    private var nfcSession: NFCNDEFReaderSession?
    private var pendingResult: FlutterResult?
    private var isReading = false
    private var isWriting = false
    private var writeData: String?
    private var eventSink: FlutterEventSink?
    private var alertMessage: String = "Hold your iPhone near an NFC business card"
    private var isDebug = true
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.swahilicard.nfc/ios", binaryMessenger: registrar.messenger())
        let instance = SwahiliNFCPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        let eventChannel = FlutterEventChannel(name: "com.swahilicard.nfc/ios/tags", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)
        
        instance.debugLog("SwahiliNFC Plugin registered")
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if #available(iOS 13.0, *) {
            switch call.method {
            case "isAvailable":
                let isAvailable = NFCNDEFReaderSession.readingAvailable
                debugLog("NFC Available: \(isAvailable)")
                result(isAvailable)
                
            case "startSession":
                guard NFCNDEFReaderSession.readingAvailable else {
                    debugLog("Error: NFC is not available on this device")
                    result(FlutterError(code: "nfc_not_available", message: "NFC is not available on this device", details: nil))
                    return
                }
                
                if let args = call.arguments as? [String: Any] {
                    isReading = args["isReading"] as? Bool ?? false
                    isWriting = args["isWriting"] as? Bool ?? false
                    alertMessage = args["alertMessage"] as? String ?? alertMessage
                }
                
                debugLog("Start session - isReading: \(isReading), isWriting: \(isWriting)")
                
                if isReading && isWriting {
                    debugLog("Error: Cannot read and write in the same session")
                    result(FlutterError(code: "invalid_args", message: "Cannot read and write in the same session", details: nil))
                    return
                }
                
                // Start NFC session
                startNFCSession()
                pendingResult = result
                
            case "stopSession":
                stopNFCSession()
                result(nil)
                
            case "getTagData":
                if pendingResult != nil {
                    debugLog("Error: Another operation is in progress")
                    result(FlutterError(code: "busy", message: "Another operation is in progress", details: nil))
                } else {
                    debugLog("Error: No tag data available")
                    result(FlutterError(code: "no_data", message: "No tag data available", details: nil))
                }
                
            case "writeTag":
                guard let args = call.arguments as? [String: Any],
                      let data = args["data"] as? String else {
                    debugLog("Error: Data is required for writing")
                    result(FlutterError(code: "invalid_args", message: "Data is required for writing", details: nil))
                    return
                }
                
                debugLog("Write data prepared: \(data.prefix(50))...")
                writeData = data
                pendingResult = result
                
                // Ensure NFC session is active
                if nfcSession == nil {
                    startNFCSession()
                }
                
            default:
                debugLog("Method not implemented: \(call.method)")
                result(FlutterMethodNotImplemented)
            }
        } else {
            // iOS version does not support NFC
            debugLog("Error: NFC is not supported on this iOS version")
            result(FlutterError(code: "unsupported_ios_version", message: "NFC is not supported on this iOS version", details: nil))
        }
    }
    
    private func startNFCSession() {
        if #available(iOS 13.0, *) {
            debugLog("Starting NFC session")
            nfcSession = NFCNDEFReaderSession(delegate: self, queue: nil, invalidateAfterFirstRead: !isWriting)
            nfcSession?.alertMessage = alertMessage
            nfcSession?.begin()
        }
    }
    
    private func stopNFCSession() {
        if #available(iOS 13.0, *) {
            debugLog("Stopping NFC session")
            nfcSession?.invalidate()
            nfcSession = nil
        }
    }
    
    // MARK: - NFCNDEFReaderSessionDelegate
    
    public func readerSession(_ session: NFCNDEFReaderSession, didInvalidateWithError error: Error) {
        // Handle session invalidation
        debugLog("Session invalidated with error: \(error.localizedDescription)")
        
        if let pendingResult = pendingResult {
            if error.localizedDescription != "Session is invalidated" {
                pendingResult(FlutterError(code: "nfc_error", message: error.localizedDescription, details: nil))
            }
            self.pendingResult = nil
        }
        
        nfcSession = nil
    }
    
    public func readerSession(_ session: NFCNDEFReaderSession, didDetectNDEFs messages: [NFCNDEFMessage]) {
        // Handle detected NDEF messages
        debugLog("Detected NDEF messages: \(messages.count)")
        handleDetectedMessages(messages)
    }
    
    @available(iOS 13.0, *)
    public func readerSession(_ session: NFCNDEFReaderSession, didDetect tags: [NFCNDEFTag]) {
        // Connect to first tag
        guard let tag = tags.first else {
            debugLog("Error: No tag detected")
            session.invalidate(errorMessage: "No tag detected")
            return
        }
        
        session.connect(to: tag) { error in
            if let error = error {
                self.debugLog("Connection failed: \(error.localizedDescription)")
                session.invalidate(errorMessage: "Connection failed: \(error.localizedDescription)")
                return
            }
            
            self.debugLog("Connected to tag")
            
            if self.isWriting && self.writeData != nil {
                // Write mode
                self.handleWriteToTag(tag, session: session)
            } else {
                // Read mode
                self.handleReadFromTag(tag, session: session)
            }
        }
    }
    
    private func handleDetectedMessages(_ messages: [NFCNDEFMessage]) {
        // Process detected messages
        guard let message = messages.first else {
            debugLog("Error: No NDEF message found")
            pendingResult?(FlutterError(code: "no_message", message: "No NDEF message found", details: nil))
            pendingResult = nil
            return
        }
        
        // Convert NDEF message to data
        let result = processNDEFMessage(message)
        debugLog("Processed NDEF message: \(result.prefix(50))...")
        
        // If this is a continuous reading session, send to event sink
        if let eventSink = eventSink {
            eventSink(result)
        }
        
        // For one-time reading, send to pending result
        if isReading {
            pendingResult?(result)
            pendingResult = nil
        }
    }
    
    @available(iOS 13.0, *)
    private func handleReadFromTag(_ tag: NFCNDEFTag, session: NFCNDEFReaderSession) {
        // Query tag for its NDEF message
        tag.queryNDEFStatus { status, capacity, error in
            if let error = error {
                self.debugLog("Failed to query tag: \(error.localizedDescription)")
                session.invalidate(errorMessage: "Failed to query tag: \(error.localizedDescription)")
                self.pendingResult?(FlutterError(code: "tag_error", message: error.localizedDescription, details: nil))
                self.pendingResult = nil
                return
            }
            
            switch status {
            case .notSupported:
                self.debugLog("Error: Tag does not support NDEF")
                session.invalidate(errorMessage: "Tag does not support NDEF")
                self.pendingResult?(FlutterError(code: "tag_error", message: "Tag does not support NDEF", details: nil))
                self.pendingResult = nil
                
            case .readOnly, .readWrite:
                // Tag supports NDEF, read message
                tag.readNDEF { message, error in
                    if let error = error {
                        self.debugLog("Failed to read NDEF: \(error.localizedDescription)")
                        session.invalidate(errorMessage: "Failed to read NDEF: \(error.localizedDescription)")
                        self.pendingResult?(FlutterError(code: "read_error", message: error.localizedDescription, details: nil))
                        self.pendingResult = nil
                        return
                    }
                    
                    if let message = message {
                        // Process NDEF message
                        let result = self.processNDEFMessage(message)
                        self.debugLog("Read successful: \(result.prefix(50))...")
                        
                        // If this is a continuous reading session, send to event sink
                        if let eventSink = self.eventSink {
                            eventSink(result)
                        }
                        
                        // For one-time reading, send to pending result
                        if self.isReading {
                            session.alertMessage = "Business card detected"
                            session.invalidate()
                            self.pendingResult?(result)
                            self.pendingResult = nil
                        }
                    } else {
                        // No message found
                        self.debugLog("Error: No NDEF message found on tag")
                        session.invalidate(errorMessage: "No NDEF message found on tag")
                        self.pendingResult?(FlutterError(code: "no_message", message: "No NDEF message found on tag", details: nil))
                        self.pendingResult = nil
                    }
                }
                
            @unknown default:
                self.debugLog("Error: Unknown tag status")
                session.invalidate(errorMessage: "Unknown tag status")
                self.pendingResult?(FlutterError(code: "unknown_status", message: "Unknown tag status", details: nil))
                self.pendingResult = nil
            }
        }
    }
    
    @available(iOS 13.0, *)
    private func handleWriteToTag(_ tag: NFCNDEFTag, session: NFCNDEFReaderSession) {
        guard let writeData = writeData else {
            debugLog("Error: No data to write")
            session.invalidate(errorMessage: "No data to write")
            pendingResult?(FlutterError(code: "no_data", message: "No data to write", details: nil))
            pendingResult = nil
            return
        }
        
        // Check tag status
        tag.queryNDEFStatus { status, capacity, error in
            if let error = error {
                self.debugLog("Failed to query tag: \(error.localizedDescription)")
                session.invalidate(errorMessage: "Failed to query tag: \(error.localizedDescription)")
                self.pendingResult?(FlutterError(code: "tag_error", message: error.localizedDescription, details: nil))
                self.pendingResult = nil
                return
            }
            
            if status == .readOnly {
                self.debugLog("Error: Tag is read-only")
                session.invalidate(errorMessage: "Tag is read-only")
                self.pendingResult?(FlutterError(code: "tag_error", message: "Tag is read-only", details: nil))
                self.pendingResult = nil
                return
            }
            
            if status == .notSupported {
                self.debugLog("Error: Tag does not support NDEF")
                session.invalidate(errorMessage: "Tag does not support NDEF")
                self.pendingResult?(FlutterError(code: "tag_error", message: "Tag does not support NDEF", details: nil))
                self.pendingResult = nil
                return
            }
            
            // Create NDEF message
            let message = self.createNDEFMessage(from: writeData)
            self.debugLog("Writing NDEF message to tag")
            
            // Write to tag
            tag.writeNDEF(message) { error in
                if let error = error {
                    self.debugLog("Write failed: \(error.localizedDescription)")
                    session.invalidate(errorMessage: "Write failed: \(error.localizedDescription)")
                    self.pendingResult?(FlutterError(code: "write_error", message: error.localizedDescription, details: nil))
                    self.pendingResult = nil
                    return
                }
                
                // Success
                self.debugLog("Write successful")
                session.alertMessage = "Write successful"
                session.invalidate()
                self.pendingResult?(true)
                self.pendingResult = nil
            }
        }
    }
    
    private func processNDEFMessage(_ message: NFCNDEFMessage) -> String {
        // Extract data from NDEF message with improved handling
        // First look for our custom MIME type
        for record in message.records {
            if let typeString = String(data: record.type, encoding: .utf8),
               typeString.contains("swahilicard") {
                
                debugLog("Found SwahiliCard MIME record")
                
                // Try various encodings for the payload
                if let payload = String(data: record.payload, encoding: .utf8) {
                    return payload
                } else if let payload = String(data: record.payload, encoding: .utf16) {
                    return payload
                } else if let payload = String(data: record.payload, encoding: .ascii) {
                    return payload
                }
            }
        }
        
        // Next, look for text records
        for record in message.records {
            if record.typeNameFormat == .nfcWellKnown {
                if let typeString = String(data: record.type, encoding: .utf8), typeString == "T" {
                    debugLog("Found TEXT record")
                    
                    // Handle text record - remove language code
                    var payload = record.payload
                    if payload.count > 0 {
                        let languageCodeLength = Int(payload[0] & 0x3F)
                        if payload.count > languageCodeLength + 1 {
                            payload = payload.advanced(by: languageCodeLength + 1)
                            
                            if let textPayload = String(data: payload, encoding: .utf8) {
                                return textPayload
                            }
                        }
                    }
                }
            }
        }
        
        // If no specific record types are found, try to get anything from the first record
        if let record = message.records.first {
            if let payload = String(data: record.payload, encoding: .utf8) {
                debugLog("Using first record with UTF-8 encoding")
                return payload
            } else if let payload = String(data: record.payload, encoding: .utf16) {
                debugLog("Using first record with UTF-16 encoding")
                return payload
            } else if let payload = String(data: record.payload, encoding: .ascii) {
                debugLog("Using first record with ASCII encoding")
                return payload
            }
        }
        
        // If all else fails, return empty JSON object
        debugLog("No readable payload found, returning empty JSON")
        return "{}"
    }
    
    @available(iOS 13.0, *)
    private func createNDEFMessage(from data: String) -> NFCNDEFMessage {
        // Create a diverse set of records for better compatibility
        var records: [NFCNDEFPayload] = []
        
        // 1. Create a MIME record with our SwahiliCard type
        let mimeType = "application/vnd.swahilicard"
        let payload = data.data(using: .utf8) ?? Data()
        
        let mimeRecord = NFCNDEFPayload(
            format: .media,
            type: mimeType.data(using: .utf8)!,
            identifier: Data(),
            payload: payload
        )
        records.append(mimeRecord)
        
        // 2. Create a text record as fallback
        if let textData = data.data(using: .utf8) {
            // Add language code (en)
            let languageCode = "en"
            let languageCodeData = languageCode.data(using: .ascii)!
            
            // Combine status byte + language code + text data
            var textPayload = Data()
            textPayload.append(UInt8(languageCodeData.count))
            textPayload.append(languageCodeData)
            textPayload.append(textData)
            
            let textRecord = NFCNDEFPayload(
                format: .nfcWellKnown,
                type: "T".data(using: .ascii)!,
                identifier: Data(),
                payload: textPayload
            )
            records.append(textRecord)
        }
        
        // 3. Add URI record with SwahiliCard website
        if let uriData = "https://swahilicard.com/".data(using: .utf8) {
            // URI record needs a prefix byte (0x00 for "http://", 0x01 for "https://", etc.)
            var uriPayload = Data([0x01]) // 0x01 for https://
            // Remove https:// from the data since it's encoded in the prefix byte
            let trimmedUri = String(uriData.dropFirst(8))
            uriPayload.append(trimmedUri)
            
            let uriRecord = NFCNDEFPayload(
                format: .nfcWellKnown,
                type: "U".data(using: .ascii)!,
                identifier: Data(),
                payload: uriPayload
            )
            records.append(uriRecord)
        }
        
        return NFCNDEFMessage(records: records)
    }
    
    // Debug logging helper
    private func debugLog(_ message: String) {
        if isDebug {
            NSLog("SwahiliNFC: \(message)")
        }
    }
}

// MARK: - FlutterStreamHandler implementation

@available(iOS 13.0, *)
extension SwahiliNFCPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        debugLog("Event sink registered for continuous reading")
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        debugLog("Event sink cancelled")
        return nil
    }
}

File: android/src/main/kotlin/com/swahilicard/nfc/SwahiliNFCPlugin.kt
package com.swahilicard.nfc

import android.app.Activity
import android.content.Intent
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.Ndef
import android.nfc.tech.NdefFormatable
import android.nfc.NdefMessage
import android.nfc.NdefRecord
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.PluginRegistry
import java.io.IOException
import java.util.concurrent.Executor
import java.util.concurrent.Executors

/** SwahiliNFCPluginKt - Kotlin implementation called by the Java wrapper */
class SwahiliNFCPluginKt: PluginRegistry.NewIntentListener {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var activity: Activity? = null
    private var nfcAdapter: NfcAdapter? = null
    private var pendingResult: Result? = null
    private var isReading = false
    private var isWriting = false
    private var writeData: String? = null
    private var tagEventSink: EventChannel.EventSink? = null
    private val executor: Executor = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())
    
    // Debug flag
    private val isDebug = true

    // Setup channels for both registration methods
    fun setupChannels(messenger: BinaryMessenger, activity: Activity?) {
        this.activity = activity
        nfcAdapter = activity?.let { NfcAdapter.getDefaultAdapter(it) }

        channel = MethodChannel(messenger, "com.swahilicard.nfc/android")
        channel.setMethodCallHandler { call, result -> onMethodCall(call, result) }
        
        eventChannel = EventChannel(messenger, "com.swahilicard.nfc/android/tags")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                tagEventSink = events
            }

            override fun onCancel(arguments: Any?) {
                tagEventSink = null
            }
        })
        
        debugLog("SwahiliNFC Plugin initialized")
    }

    fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        // This method is called by the Java wrapper
    }

    fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "isAvailable" -> {
                val isAvailable = nfcAdapter != null && nfcAdapter!!.isEnabled
                debugLog("NFC Available: $isAvailable")
                result.success(isAvailable)
            }
            "startSession" -> {
                isReading = call.argument<Boolean>("isReading") ?: false
                isWriting = call.argument<Boolean>("isWriting") ?: false
                
                debugLog("Start session - isReading: $isReading, isWriting: $isWriting")
                
                if (isWriting && isReading) {
                    debugLog("Error: Cannot read and write in the same session")
                    result.error(
                        "invalid_args",
                        "Cannot read and write in the same session",
                        null
                    )
                    return
                }
                
                if (nfcAdapter == null || !nfcAdapter!!.isEnabled) {
                    debugLog("Error: NFC not available or disabled")
                    result.error(
                        "nfc_not_available",
                        "NFC is not available or disabled",
                        null
                    )
                    return
                }
                
                // Enable foreground dispatch
                enableForegroundDispatch()
                
                // For read sessions, we'll wait for a tag
                if (isReading) {
                    pendingResult = result
                    debugLog("Read session started - waiting for tag")
                } else {
                    // For write sessions, return success immediately
                    debugLog("Write session started")
                    result.success(null)
                }
            }
            "stopSession" -> {
                disableForegroundDispatch()
                isReading = false
                isWriting = false
                writeData = null
                pendingResult = null
                debugLog("Session stopped")
                result.success(null)
            }
            "readTag" -> {
                // This is handled in onNewIntent for Android
                debugLog("Error: No tag detected. Call startSession first.")
                result.error(
                    "invalid_state",
                    "No tag detected. Call startSession first.",
                    null
                )
            }
            "writeTag" -> {
                val data = call.argument<Any>("data")
                if (data == null) {
                    debugLog("Error: Data is required for writing")
                    result.error(
                        "invalid_args",
                        "Data is required for writing",
                        null
                    )
                    return
                }
                
                // Convert the data to a JSON string regardless of whether it's a Map or String
                val jsonData = when (data) {
                    is Map<*, *> -> {
                        try {
                            // Convert map to JSON string
                            debugLog("Converting Map to JSON for writing")
                            org.json.JSONObject(data).toString()
                        } catch (e: Exception) {
                            debugLog("Error converting data to JSON: ${e.message}")
                            result.error(
                                "conversion_error",
                                "Failed to convert data to JSON: ${e.message}",
                                null
                            )
                            return
                        }
                    }
                    is String -> {
                        debugLog("Using string data for writing: ${data.take(50)}...")
                        data // Already a string
                    }
                    else -> {
                        debugLog("Error: Data must be a Map or String")
                        result.error(
                            "invalid_args",
                            "Data must be a Map or String",
                            null
                        )
                        return
                    }
                }
                
                writeData = jsonData
                pendingResult = result
                debugLog("Write data prepared, waiting for tag")
            }
            else -> {
                debugLog("Method not implemented: ${call.method}")
                result.notImplemented()
            }
        }
    }

    fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // This method is called by the Java wrapper
    }

    fun onAttachedToActivity(binding: io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding) {
        activity = binding.activity
        binding.addOnNewIntentListener(this)
        nfcAdapter = NfcAdapter.getDefaultAdapter(activity)
        debugLog("Plugin attached to activity")
    }

    fun onDetachedFromActivityForConfigChanges() {
        disableForegroundDispatch()
        debugLog("Plugin detached from activity for config changes")
    }

    fun onReattachedToActivityForConfigChanges(binding: io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding) {
        activity = binding.activity
        binding.addOnNewIntentListener(this)
        nfcAdapter = NfcAdapter.getDefaultAdapter(activity)
        debugLog("Plugin reattached to activity")
    }

    fun onDetachedFromActivity() {
        disableForegroundDispatch()
        activity = null
        debugLog("Plugin detached from activity")
    }

    override fun onNewIntent(intent: Intent): Boolean {
        // Handle new NFC intent
        if (NfcAdapter.ACTION_NDEF_DISCOVERED == intent.action ||
            NfcAdapter.ACTION_TECH_DISCOVERED == intent.action ||
            NfcAdapter.ACTION_TAG_DISCOVERED == intent.action) {
            
            debugLog("NFC tag detected: ${intent.action}")
            
            val tag = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                intent.getParcelableExtra(NfcAdapter.EXTRA_TAG, Tag::class.java)
            } else {
                @Suppress("DEPRECATION")
                intent.getParcelableExtra(NfcAdapter.EXTRA_TAG)
            }
            
            if (tag != null) {
                if (isReading) {
                    debugLog("Processing tag for reading")
                    handleReadTag(tag)
                } else if (isWriting) {
                    debugLog("Processing tag for writing")
                    handleWriteTag(tag)
                }
                return true
            } else {
                debugLog("Tag is null")
            }
        }
        return false
    }

    private fun handleReadTag(tag: Tag) {
        executor.execute {
            try {
                val ndef = Ndef.get(tag)
                if (ndef != null) {
                    ndef.connect()
                    debugLog("Connected to NDEF tag")
                    
                    val ndefMessage = ndef.cachedNdefMessage
                    if (ndefMessage != null) {
                        debugLog("NDEF message found with ${ndefMessage.records.size} records")
                        
                        // First look for our specific mime type
                        val mimeRecord = ndefMessage.records.find { record ->
                            record.tnf == NdefRecord.TNF_MIME_MEDIA &&
                            String(record.type).contains("swahilicard")
                        }
                        
                        val result = if (mimeRecord != null) {
                            // Found our specific record
                            debugLog("Found SwahiliCard MIME record")
                            String(mimeRecord.payload)
                        } else {
                            // Fallback: check for well-known text record
                            val textRecord = ndefMessage.records.find { record ->
                                record.tnf == NdefRecord.TNF_WELL_KNOWN &&
                                record.type.contentEquals(NdefRecord.RTD_TEXT)
                            }
                            
                            if (textRecord != null) {
                                // Process text record - need to skip language code
                                debugLog("Found TEXT record")
                                val payload = textRecord.payload
                                val languageCodeLength = payload[0].toInt() and 0x3F
                                String(payload, languageCodeLength + 1, payload.size - languageCodeLength - 1)
                            } else {
                                // Last resort - just use first record
                                debugLog("No special record found, using first record")
                                val firstRecord = ndefMessage.records.firstOrNull()
                                if (firstRecord != null) {
                                    try {
                                        String(firstRecord.payload)
                                    } catch (e: Exception) {
                                        debugLog("Error converting payload to string: ${e.message}")
                                        "{}" // Empty JSON as fallback
                                    }
                                } else {
                                    debugLog("No records in NDEF message")
                                    "{}"
                                }
                            }
                        }
                        
                        debugLog("Read result (first 50 chars): ${result.take(50)}...")
                        
                        ndef.close()
                        
                        // Notify event channel for continuous reading
                        if (tagEventSink != null) {
                            mainHandler.post {
                                tagEventSink?.success(result)
                            }
                        }
                        
                        // Return result for one-time reading
                        mainHandler.post {
                            pendingResult?.success(result)
                            pendingResult = null
                        }
                    } else {
                        // No NDEF message on tag
                        debugLog("No NDEF message on tag")
                        val emptyResult = "{}"
                        
                        ndef.close()
                        
                        mainHandler.post {
                            pendingResult?.success(emptyResult)
                            pendingResult = null
                        }
                    }
                } else {
                    // Tag doesn't support NDEF
                    val error = "Tag doesn't support NDEF"
                    debugLog(error)
                    
                    mainHandler.post {
                        pendingResult?.error("tag_error", error, null)
                        pendingResult = null
                    }
                }
            } catch (e: IOException) {
                val error = "Error reading from tag: ${e.message}"
                debugLog(error)
                
                mainHandler.post {
                    pendingResult?.error("io_error", error, null)
                    pendingResult = null
                }
            } catch (e: Exception) {
                val error = "Error: ${e.message}"
                debugLog(error)
                
                mainHandler.post {
                    pendingResult?.error("unknown_error", error, null)
                    pendingResult = null
                }
            }
        }
    }

    private fun handleWriteTag(tag: Tag) {
    if (writeData == null) {
        debugLog("Error: No data to write")
        mainHandler.post {
            pendingResult?.error("invalid_state", "No data to write", null)
            pendingResult = null
        }
        return
    }
    
    debugLog("Writing data to tag (first 50 chars): ${writeData!!.take(50)}...")
    
    executor.execute {
        try {
            val ndef = Ndef.get(tag)
            
            if (ndef != null) {
                // Tag supports NDEF, write directly
                try {
                    ndef.connect()
                    debugLog("Connected to NDEF tag")
                    
                    // Check if tag is writable
                    if (!ndef.isWritable) {
                        debugLog("Error: Tag is not writable")
                        mainHandler.post {
                            pendingResult?.error("tag_error", "Tag is not writable", null)
                            pendingResult = null
                        }
                        return@execute
                    }
                    
                    // Check tag capacity - this is crucial
                    val tagSize = ndef.maxSize
                    debugLog("Tag capacity: $tagSize bytes")
                    
                    // Create NDEF message
                    val ndefMessage = createNDEFMessage(writeData!!)
                    val messageSize = ndefMessage.byteArrayLength
                    debugLog("Message size: $messageSize bytes")
                    
                    if (messageSize > tagSize) {
                        val errorMsg = "Message too large ($messageSize bytes) for tag capacity ($tagSize bytes)"
                        debugLog("Error: $errorMsg")
                        mainHandler.post {
                            pendingResult?.error(
                                "size_error",
                                errorMsg,
                                null
                            )
                            pendingResult = null
                        }
                        return@execute
                    }
                    
                    // Try writing with explicit error handling
                    debugLog("Writing NDEF message to tag")
                    try {
                        ndef.writeNdefMessage(ndefMessage)
                        debugLog("Write successful")
                        
                        // Wait briefly to ensure write is complete
                        Thread.sleep(50)
                        
                        mainHandler.post {
                            pendingResult?.success(true)
                            pendingResult = null
                        }
                    } catch (e: Exception) {
                        val errorMsg = if (e.message != null) e.message else "Tag write failed - possibly moved too quickly"
                        debugLog("Write error: $errorMsg")
                        debugLog("Exception type: ${e.javaClass.simpleName}")
                        e.printStackTrace()
                        
                        mainHandler.post {
                            pendingResult?.error(
                                "write_error",
                                errorMsg,
                                null
                            )
                            pendingResult = null
                        }
                    } finally {
                        try {
                            ndef.close()
                        } catch (e: Exception) {
                            debugLog("Error closing NDEF connection: ${e.message}")
                        }
                    }
                } catch (e: Exception) {
                    val errorMsg = e.message ?: "Unknown error connecting to tag"
                    debugLog("Connection error: $errorMsg")
                    debugLog("Exception type: ${e.javaClass.simpleName}")
                    e.printStackTrace()
                    
                    mainHandler.post {
                        pendingResult?.error("connection_error", errorMsg, null)
                        pendingResult = null
                    }
                    
                    try {
                        ndef.close()
                    } catch (closeEx: Exception) {
                        // Ignore close errors
                    }
                }
            } else {
                // Try to format tag
                debugLog("Tag doesn't support NDEF, trying to format")
                val ndefFormatable = NdefFormatable.get(tag)
                
                if (ndefFormatable != null) {
                    try {
                        ndefFormatable.connect()
                        debugLog("Connected to formattable tag")
                        
                        // Create NDEF message
                        val ndefMessage = createNDEFMessage(writeData!!)
                        
                        debugLog("Formatting tag and writing NDEF message")
                        ndefFormatable.format(ndefMessage)
                        debugLog("Format and write successful")
                        
                        // Wait briefly to ensure write is complete
                        Thread.sleep(50)
                        
                        mainHandler.post {
                            pendingResult?.success(true)
                            pendingResult = null
                        }
                    } catch (e: Exception) {
                        val errorMsg = e.message ?: "Unknown formatting error"
                        debugLog("Format error: $errorMsg")
                        debugLog("Exception type: ${e.javaClass.simpleName}")
                        e.printStackTrace()
                        
                        mainHandler.post {
                            pendingResult?.error(
                                "format_error",
                                errorMsg,
                                null
                            )
                            pendingResult = null
                        }
                    } finally {
                        try {
                            ndefFormatable.close()
                        } catch (e: Exception) {
                            // Ignore close errors
                        }
                    }
                } else {
                    val errorMsg = "Tag doesn't support NDEF and cannot be formatted"
                    debugLog("Error: $errorMsg")
                    mainHandler.post {
                        pendingResult?.error(
                            "tag_error",
                            errorMsg,
                            null
                        )
                        pendingResult = null
                    }
                }
            }
        } catch (e: Exception) {
            val errorMsg = e.message ?: "Unknown error in tag processing"
            debugLog("Processing error: $errorMsg")
            debugLog("Exception type: ${e.javaClass.simpleName}")
            e.printStackTrace()
            
            mainHandler.post {
                pendingResult?.error("processing_error", errorMsg, null)
                pendingResult = null
            }
        }
    }
}

    private fun enableForegroundDispatch() {
        try {
            val currentActivity = activity ?: return
            if (nfcAdapter != null && nfcAdapter!!.isEnabled) {
                val intent = Intent(currentActivity, currentActivity.javaClass).apply {
                    addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
                }
                val pendingIntent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    android.app.PendingIntent.getActivity(
                        currentActivity,
                        0,
                        intent,
                        android.app.PendingIntent.FLAG_MUTABLE
                    )
                } else {
                    @Suppress("DEPRECATION")
                    android.app.PendingIntent.getActivity(
                        currentActivity,
                        0,
                        intent,
                        0
                    )
                }
                
                nfcAdapter!!.enableForegroundDispatch(currentActivity, pendingIntent, null, null)
                debugLog("Foreground dispatch enabled")
            }
        } catch (e: Exception) {
            debugLog("Error enabling foreground dispatch: ${e.message}")
            e.printStackTrace()
        }
    }

    private fun disableForegroundDispatch() {
        try {
            val currentActivity = activity ?: return
            if (nfcAdapter != null && nfcAdapter!!.isEnabled) {
                nfcAdapter!!.disableForegroundDispatch(currentActivity)
                debugLog("Foreground dispatch disabled")
            }
        } catch (e: Exception) {
            debugLog("Error disabling foreground dispatch: ${e.message}")
            e.printStackTrace()
        }
    }

    private fun createNDEFMessage(data: String): NdefMessage {
        // For SwahiliCard, use our custom MIME type
        val mimeType = "application/vnd.swahilicard"
        val mimeRecord = NdefRecord.createMime(
            mimeType,
            data.toByteArray()
        )
        
        // Also create a text record as fallback for apps that don't understand our MIME type
        val languageCode = "en"
        val textBytes = data.toByteArray()
        val textPayload = ByteArray(1 + languageCode.length + textBytes.size)
        textPayload[0] = languageCode.length.toByte()
        System.arraycopy(languageCode.toByteArray(), 0, textPayload, 1, languageCode.length)
        System.arraycopy(textBytes, 0, textPayload, 1 + languageCode.length, textBytes.size)
        val textRecord = NdefRecord(
            NdefRecord.TNF_WELL_KNOWN,
            NdefRecord.RTD_TEXT,
            ByteArray(0),
            textPayload
        )
        
        // Create a URI record with a web link as additional fallback
        val uriRecord = NdefRecord.createUri("https://swahilicard.com/")
        
        // Return message with all records
        return NdefMessage(arrayOf(mimeRecord, textRecord, uriRecord))
    }
    
    // Debug logging helper
    private fun debugLog(message: String) {
        if (isDebug) {
            Log.d("SwahiliNFC", message)
        }
    }
}